src/bloom.cpp

```diff
@@ -34,12 +34,12 @@ CBloomFilter::CBloomFilter(unsigned int nElements, double nFPRate, unsigned int
      * Again, we ignore filter parameters which will create a bloom filter with more hash functions than the protocol limits
      * See https://en.wikipedia.org/wiki/Bloom_filter for an explanation of these formulas
      */
-    isFull(false),
     isEmpty(false),
     nHashFuncs(min((unsigned int)(vData.size() * 8 / nElements * LN2), MAX_HASH_FUNCS)),
     nTweak(nTweakIn),
     nFlags(nFlagsIn)
 {
+    isFull = vData.size() == 0;
```

Comment:
Excellent points and good catch regarding clear().

> I think that it makes a lot more sense to require that the minimum vData size something >= 1. Perhaps 8 which would be a negligible # of bytes on the wire and be useful for degenerate cases where we have < 100 elements. This ensures that we don't have DIV0s and OOB bombs hiding in the code, protected by a non-intuitive boolean, and therefore simplifies code analysis. What are the drawbacks of this approach?

I imagine that it might be harder to guarantee that users of the bloom filter can always work with a bloom filter of truly more than 1 element? I have no strong opinion on this however.


A couple further points:
When we make the filter more strict, there's potential interoperability issues with Core whenever a command works with a bloom filter. Do we risk something when we ban Core nodes for supplying us with bloom filters that violate BU's then-stricter constraints?
I think you'd still likely want to ban peers that send you trashy blooms, and our harder constraints would also be a likely indication of attack. 

There's the serialization/deserialization angle, too: Right now, an empty bloom filter can be deserialized and will then be in a broken state which will OOB if UpdateEmptyFull is not called. If we disallow empty bloom filters, we need to figure out what to do with the deserialization of an empty filter.

Should we simply add this condition to IsWithinSizeConstraints() and will that be enough to avoid  anyone dealing with this filter?

I don't really think so: I think it would be great if *any* bloom filter, constructed in any way (either through deserialization or one of the supplied constructors) won't have any modes of hard failure like OOB segfaulting or throwing an exception.


I suggest that maybe we want to always 'sanitize' the filter after construction or deserialization but set some kind of internal flag if the values supplied are bogus.

I think that would be the most robust approach: The object always is in a state where it can be used in any way, however there's a 'isSane()' call which will only return true if the filter wasn't constructed or deserialized with a bogus size range or other shenanigans.

Making it robust also means then, though, that setting nElements = 0 would be overriden to mean nElements=1 internally (plus setting the insane flag), and the filter would ignore the user input on this. There might be unforeseen issues here as well and it might be good to think about it, but right now I can't see any.

(IMO this approach of making objects as robust as possible is in general a good thing in general for bitcoind QA)

I can rework the patch as soon as we think we have a good idea on what the implementation should be - or if you think this 'more robust implementation with isSane()' approach sounds like the right way to go.